cpp/tests/utilities/base_fixture.hpp

```diff
@@ -165,43 +165,36 @@ class TempDirTestEnvironment : public ::testing::Environment {
   }
 };
 
-
-/**---------------------------------------------------------------------------*
- * @brief Test environment that initializes the default rmm memory resource.
+/**
+ * @brief Creates and sets the default memory resource for the
+ * unit test environment.
  * 
- * Required for tests programs that use rmm. It is recommended to include
- * `CUDF_TEST_PROGRAM_MAIN()` in a code file instead of directly instantiating 
- * an object of this type.
- *---------------------------------------------------------------------------**/
-class RmmTestEnvironment : public ::testing::Environment {
-/**---------------------------------------------------------------------------*
- * @brief String representing which RMM allocation mode is to be used.
- *
+ * The returned resource instance must be kept alive for the duration of
+ * the tests. Attaching the resource to a TestEnvironment causes
+ * issues since the environment objects are not destroyed until
+ * after the runtime is shutdown.
  * 
- *---------------------------------------------------------------------------**/
-  std::unique_ptr<rmm::mr::device_memory_resource> rmm_resource{};
-public:
-  /**
-   * @brief Sets the default RMM memory resource based on the input string.
-   * 
-   * @param allocation_mode Represents the type of the memory resource to be 
-   * used as a default resource. Valid values are 'cuda', 'pool' and 'managed'.
-   * 
-   * @throws cudf::logic_error if passed mode value is invalid.
-   */
-  RmmTestEnvironment(std::string const& allocation_mode) {
+ * @throw cudf::logic_error if the `allocation_mode` is unsupported.
+ *
+ * @param allocation_mode String identifies which resource type.
+ *        Accepted types are "pool", "cuda", and "managed" only.
+ * @return Memory resource instance
+ */
+inline auto create_memory_resource(std::string const& allocation_mode)
+{
+    std::unique_ptr<rmm::mr::device_memory_resource> rmm_resource{};
     if (allocation_mode == "cuda")
       rmm_resource.reset(new rmm::mr::cuda_memory_resource());
     else if (allocation_mode == "pool")
       rmm_resource.reset(new rmm::mr::cnmem_memory_resource());
     else if (allocation_mode == "managed")
       rmm_resource.reset(new rmm::mr::managed_memory_resource());
-    else 
+    else
       CUDF_FAIL("Invalid RMM allocation mode");
 
     rmm::mr::set_default_resource(rmm_resource.get());
```

Comment:
Maybe move `set_default_resource` to `main` after calling `create_memory_resource`? Then I suppose the branches above could be changed to `return make_unique<...>(...)`

Reply:
I don't think `make_unique` is useful here because the pointer type is different from the owned object type.

Reply:
Ah, I think the return type would need to explicitly be `std::unique_ptr<rmm::mr::device_memory_resource>` rather than `auto` in this case, but I believe the `unique_ptr`s will implicitly cast to the parent type here.